utils/stream/stream.go

```diff
@@ -0,0 +1,189 @@
+// Copyright 2018 the Service Broker Project Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package stream
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	yaml "gopkg.in/yaml.v2"
+)
+
+type Source func() (io.ReadCloser, error)
+type Dest func() (io.WriteCloser, error)
+
+// Copy copies data from a source stream to a destination stream.
+func Copy(src Source, dest Dest) error {
+	readCloser, err := src()
+	if err != nil {
+		return fmt.Errorf("copy couldn't open source: %v", err)
+	}
+	defer readCloser.Close()
+
+	writeCloser, err := dest()
+	if err != nil {
+		return fmt.Errorf("copy couldn't open destination: %v", err)
+	}
+	defer writeCloser.Close()
+
+	if _, err := io.Copy(writeCloser, readCloser); err != nil {
+		return fmt.Errorf("copy couldn't copy data: %v", err)
+	}
+
+	if err := readCloser.Close(); err != nil {
```

Comment:
I'm concerned about the fact that Close() is called twice on both the writer and the reader. Here's what the docs have to say:
"The behavior of Close after the first call is undefined. Specific implementations may document their own behavior."

Reply:
Just noticed the use of io.NoOpCloser below. I make make mention of that assumption in the method header for future maintainers.

Reply:
You were right the first time, I'm adding a multi-closer to handle this case that will guarantee at-most-once close semantics.

I found lots of fun discussion about adding a `OnceCloser()` to the stdlib, but as always it ended up in the state of "this is only 10 lines, just add it to your own project"...ugh.